packages/tools/examples/segmentBidirectionalTool/findLargestBidirectional.ts

```diff
@@ -0,0 +1,218 @@
+
+  console.time('contour');
+  const contours = generateContourSetsFromLabelmap({
+    segmentations,
+    cornerstoneCache: cache,
+    cornerstoneToolsEnums: cornerstoneTools.Enums,
+    vtkUtils,
+  });
+  console.timeEnd('contour');
+
+  if (!contours?.length || !contours[0].sliceContours.length) {
+    return;
+  }
+
+  const { representationData, segments = [0, 1] } = segmentations;
+  const { Labelmap } = cornerstoneTools.Enums.SegmentationRepresentations;
+  const { volumeId: segVolumeId } = representationData[Labelmap];
+
+  const segmentIndex = segments.findIndex((it) => !!it);
+  if (segmentIndex === -1) {
+    return;
+  }
+  segments[segmentIndex].segmentIndex = segmentIndex;
+  return findLargestBidirectional(
+    contours[0],
+    segVolumeId,
+    segments[segmentIndex]
+  );
+}
+
+function createIsInSegment(segVolumeId: string, segment: number) {
+  // Get segmentation volume
+  const vol = cache.getVolume(segVolumeId);
+  if (!vol) {
+    console.warn(`No volume found for ${segVolumeId}`);
+    return;
+  }
+
+  const segData = vol.imageData.getPointData().getScalars().getData();
+  const width = vol.dimensions[0];
+  const pixelsPerSlice = width * vol.dimensions[1];
+
+  return (pointI, pointJ) => {
+    const point = vec3.add(vec3.create(), pointI, pointJ).map((it) => it / 2);
+    const ijk = vol.imageData.worldToIndex(point as vec3).map(Math.round);
+    const [i, j, k] = ijk;
+    const index = i + j * width + k * pixelsPerSlice;
+    const value = segData[index];
+    return value === segment;
+  };
+}
+
+function findLargestBidirectional(contours, segVolumeId, segment) {
+  const { sliceContours } = contours;
+  const { segmentIndex } = segment;
+  let maxBidirectional;
+  const isInSegment = createIsInSegment(segVolumeId, segmentIndex);
+  for (const sliceContour of sliceContours) {
+    const startTimeMs = Date.now();
+    const bidirectional = createBidirectionalForSlice(
+      sliceContour,
+      maxBidirectional,
+      isInSegment
+    );
+    const durationMs = Date.now() - startTimeMs;
+    if (!bidirectional) {
+      if (durationMs > 250) {
+        console.log('Skipped slice bidirectional took', durationMs);
+      }
+      continue;
+    }
+    maxBidirectional = bidirectional;
+    if (durationMs > 250) {
+      console.log('Slice bidirectional calculation took', durationMs);
+    }
+  }
+  if (maxBidirectional) {
+    Object.assign(maxBidirectional, segment);
+  }
+  return maxBidirectional;
+}
+
+const lengthSq = (point) =>
+  point[0] * point[0] + point[1] * point[1] + point[2] * point[2];
+
+const distanceSq = (p1, p2) => lengthSq(vec3.sub(vec3.create(), p1, p2));
+
+/**
+ * Determines if there is a point in points other than pointI or pointJ which
+ * is on the line between I and J.
+ */
+function isCrossing(points, pointI, pointJ, distance) {
+  for (const point of points) {
+    if (point === pointI || point === pointJ) {
```

Comment:
This will never be true, since you are comparing objects/arrays

Reply:
Not so, I'm compare two objects to see that the object doesn't get compared to itself.  Hmm, can actually start with position of pointI + 1 - will skip 50% of the comparisons on average, definitely worth doing.